commit b6b76dd62c56f257cdd30900ed22668c42a74030
Author: Masami Hiramatsu <mhiramat@kernel.org>
Date:   Mon Mar 12 19:00:49 2018 +0900

    error-injection: Fix to prohibit jump optimization
    
    Since the kprobe which was optimized by jump can not change
    the execution path, the kprobe for error-injection must not
    be optimized. To prohibit it, set a dummy post-handler as
    officially stated in Documentation/kprobes.txt.
    
    Fixes: 4b1a29a7f542 ("error-injection: Support fault injection framework")
    Signed-off-by: Masami Hiramatsu <mhiramat@kernel.org>
    Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

diff --git a/kernel/fail_function.c b/kernel/fail_function.c
index 21b0122cb39c..1d5632d8bbcc 100644
--- a/kernel/fail_function.c
+++ b/kernel/fail_function.c
@@ -14,6 +14,15 @@
 
 static int fei_kprobe_handler(struct kprobe *kp, struct pt_regs *regs);
 
+static void fei_post_handler(struct kprobe *kp, struct pt_regs *regs,
+			     unsigned long flags)
+{
+	/*
+	 * A dummy post handler is required to prohibit optimizing, because
+	 * jump optimization does not support execution path overriding.
+	 */
+}
+
 struct fei_attr {
 	struct list_head list;
 	struct kprobe kp;
@@ -56,6 +65,7 @@ static struct fei_attr *fei_attr_new(const char *sym, unsigned long addr)
 			return NULL;
 		}
 		attr->kp.pre_handler = fei_kprobe_handler;
+		attr->kp.post_handler = fei_post_handler;
 		attr->retval = adjust_error_retval(addr, 0);
 		INIT_LIST_HEAD(&attr->list);
 	}